core/src/mast/node/mod.rs

```diff
@@ -106,6 +106,12 @@ impl MastNode {
 // ------------------------------------------------------------------------------------------------
 /// Public accessors
 impl MastNode {
+    /// Returns true if this node is a local node (i.e., not an External or Dyn node).
+    pub fn is_local(&self) -> bool {
+        !matches!(self, MastNode::External(_) | MastNode::Dyn)
```

Comment:
Why do we also include `Dyn` in this? While I see that all `Dyn`s have the same hash, and that would cause problems in the `MastForestStore`, I don't see `Dyn` as being synonymous with "not local".

Wouldn't it be better to only return `External` here, and in the `MastForestStore`, add a check so as to not add an entry for the `Dyn` node hash?

Reply:
Good point! One thing I wanted to achieve is to have a single way of getting all procedure digests which are actually contained in a MAST forest. So, the method which is really relevant is `MastForest::local_procedure_digests()` and that one relies on this method.

I agree that `Dyn` is not really "non-local" - but I also didn't want to include a procedure which is a single `Dyn` node in that list so that we wouldn't need to do extra checks in `MemMastForestStore`.

But thinking more about this, maybe not allowing procedures which are a single `Dyn` node in the first place is a better solution.

Reply:
Actually, thinking more about this - not allowing procedures which are just `Dyn` nodes is probably not a good idea. For example, it would make the following program impossible:
```
being
   dynexec
end
```
But I'm not sure I see a reason to make a program like this illegal.

Reply:
I've implemented the suggestion in the latest commit. Thinking more about this, it doesn't seem like having procedures which  are defined by a single `Dyn` node would be a problem for a MAST forest store.

Reply:
Also people *might* want 

```
export.f
  dynexec
end
```

for API compatibility reasons? Backing up, what was the problem with allowing these procedures in the `MastForestStore` in the first place? Seems like you could return any MAST forest that contains such a node, and things should work fine?

Reply:
I'm assuming the actual problem we're having is that we may end up overwriting the "real" definition of the external procedure in the store (or perhaps we ignore insertions of the real definition because an `External` already exists), and so end up in an infinite loop of trying to get the procedure for an `External` node - or at least that's what it looks like to me.

So it's not so much that the procedures can't be allowed, so much as their definitions can't be used to resolve look ups for an `External`. We could probably catch this with an assertion by checking if we get back an `External` node with the same digest as the one we asked for, but that would just help us catch things like this in the future, and isn't a fix for the root cause. I think the fix for _that_ is to always prefer the "real" node, regardless of what order `MastForest`s are added to the store.

Reply:
> I'm assuming the actual problem we're having is that we may end up overwriting the "real" definition of the external procedure in the store (or perhaps we ignore insertions of the real definition because an `External` already exists), and so end up in an infinite loop of trying to get the procedure for an `External` node - or at least that's what it looks like to me.

Yes - that's exactly what's happening.

> So it's not so much that the procedures can't be allowed, so much as their definitions can't be used to resolve look ups for an `External`. We could probably catch this with an assertion by checking if we get back an `External` node with the same digest as the one we asked for, but that would just help us catch things like this in the future, and isn't a fix for the root cause. I think the fix for _that_ is to always prefer the "real" node, regardless of what order `MastForest`s are added to the store.

The fix in this PR should make everything work fine for `MemMastForestStore` - but I agree with the suggestion that if we get back an `External` node on look-up we should probably just error out because things are likely to go into an infinite loop.